mitmproxy/tools/web/app.py

```diff
@@ -324,6 +327,72 @@ def broadcast(cls, **kwargs):
 class ClientConnection(WebSocketEventBroadcaster):
     connections: ClassVar[set] = set()
 
+    def __init__(self, application: Application, request, **kwargs):
+        super().__init__(application, request, **kwargs)
+        self.application: Application = application
+        self.filters: dict[str, flowfilter.TFilter] = {}  # Instance-level filters
+
+    def send_flow(self, cmd: str, f: mitmproxy.flow.Flow):
```

Comment:
I think we can make this work as follows:

```python3
    @classmethod
    def broadcast_flow(cls, cmd: str, f: mitmproxy.flow.Flow) -> None:
        flow_json = flow_to_json(f)
        for conn in cls.connections:
            conn._broadcast_flow(cmd, flow_json)

    def _broadcast_flow(self, cmd: str, f: Any) -> None:
        # <current contents of send_flow>
```

Reply:
yes but with a small change:
```
def _broadcast_flow(self, cmd: str, f: mitmproxy.flow.Flow) -> None:
        matches = {
            expr.pattern: bool(expr(f))
            for expr in self.filters.values()
        }
```
so we can use `expr(f)` because f is a `Flow`, otherwise it would be a `dict` and it doesn't work!

Reply:
Ah, great catch. But let's pass both the dict and the flow as arguments then. `flow_to_json` is doing a bunch of work which would be nice to reuse across clients. :)

Reply:
makes sense! done :)